### Question 1

In [26]:
import pandas as pd

# Use a raw string for the file path to handle backslashes and spaces
file_path = r"C:\Users\shawn\OneDrive - The University of Auckland\Python Exam files\bank_datset.txt"

# Try reading the file
bank_loans = pd.read_csv(file_path, skiprows=23)

# Check if it worked
print(bank_loans.info())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   CustomerNo       418 non-null    int64 
 1   ApplicationDate  418 non-null    object
 2   LoanPurpose      418 non-null    object
 3   Checking         418 non-null    object
 4   Savings          418 non-null    object
 5   MonthsCustomer   418 non-null    int64 
 6   MonthsEmployed   418 non-null    int64 
 7   EducationLevel   418 non-null    object
 8   Gender           418 non-null    object
 9   MaritalStatus    418 non-null    object
 10  Age              418 non-null    int64 
 11  Housing          418 non-null    object
 12  Years            418 non-null    int64 
 13  Job              418 non-null    object
 14  CreditRisk       418 non-null    object
dtypes: int64(5), object(10)
memory usage: 49.1+ KB
None


In [27]:
#  Use appropriate Python command(s) to inspect bank_loans 

# Q1.1: How many loan applications are there in total?
total_applications = len(bank_loans)
print("Total number of loan applications:", total_applications)

# Q1.2: How many columns are there in total?
total_columns = bank_loans.shape[1]
print("Total number of columns:", total_columns)

# Inspect column data types
print("\nColumn Data Types:")
print(bank_loans.dtypes)

# Identify columns with integer values
int_columns = bank_loans.select_dtypes(include='int64').columns.tolist()
print("\nColumns containing integer values:", int_columns)

# Identify columns with string/text values
string_columns = bank_loans.select_dtypes(include='object').columns.tolist()
print("\nColumns containing string/text values:", string_columns)

Total number of loan applications: 418
Total number of columns: 15

Column Data Types:
CustomerNo          int64
ApplicationDate    object
LoanPurpose        object
Checking           object
Savings            object
MonthsCustomer      int64
MonthsEmployed      int64
EducationLevel     object
Gender             object
MaritalStatus      object
Age                 int64
Housing            object
Years               int64
Job                object
CreditRisk         object
dtype: object

Columns containing integer values: ['CustomerNo', 'MonthsCustomer', 'MonthsEmployed', 'Age', 'Years']

Columns containing string/text values: ['ApplicationDate', 'LoanPurpose', 'Checking', 'Savings', 'EducationLevel', 'Gender', 'MaritalStatus', 'Housing', 'Job', 'CreditRisk']


#### **Question 1** Answers

1. **How many loan applications are there in total?**  
   There are **418** loan applications in total.

2. **How many columns are there in total? Which column(s) contain integer values? Which column(s) contain string/text values?**  
   There are **15** columns in total.
   The following columns contain integer values:  
   - `CustomerNo`  
   - `MonthsCustomer`  
   - `MonthsEmployed`  
   - `Age`  
   - `Years`
   
   The following columns contain string/text values:  
   - `ApplicationDate`  
   - `LoanPurpose`  
   - `Checking`  
   - `Savings`  
   - `EducationLevel`  
   - `Gender`  
   - `MaritalStatus`  
   - `Housing`  
   - `Job`  
   - `CreditRisk`

### Question 2

In [28]:
# Add a new column called TotalBalance

# Use cleaned temporary values to calculate balance
checking_values = bank_loans['Checking'].replace(r'[\$,]', '', regex=True).astype(float)
savings_values = bank_loans['Savings'].replace(r'[\$,]', '', regex=True).astype(float)

bank_loans['TotalBalance'] = checking_values + savings_values


In [29]:
# List the customer information of the top 15 customers who have the lowest total balances

# Sort by TotalBalance ascending, then by CustomerNo ascending
lowest_balances = bank_loans.sort_values(by=['TotalBalance', 'CustomerNo']).reset_index(drop=True)

# Display top 15 customers with lowest total balances
print("Top 15 customers with the lowest TotalBalance:")
print(lowest_balances[['CustomerNo', 'TotalBalance']].head(15))

Top 15 customers with the lowest TotalBalance:
    CustomerNo  TotalBalance
0            6           0.0
1           26           0.0
2           30           0.0
3           59           0.0
4           74           0.0
5          134           0.0
6          138           0.0
7          195           0.0
8          286           0.0
9          291           0.0
10         411           0.0
11         418           0.0
12         320         102.0
13          90         104.0
14         139         104.0


In [30]:
# Q2.1: Get customer number for the fifth observation
fifth_cust_no = lowest_balances.loc[4, 'CustomerNo']
print("\nCustomer number for the fifth observation:", fifth_cust_no)

# Q2.2: Difference between lowest and 15th lowest TotalBalance
min_balance = lowest_balances.loc[0, 'TotalBalance']
fifteenth_balance = lowest_balances.loc[14, 'TotalBalance']
balance_difference = fifteenth_balance - min_balance
print("Difference between lowest and 15th-lowest total balances:", balance_difference)


Customer number for the fifth observation: 74
Difference between lowest and 15th-lowest total balances: 104.0


#### **Question 2** Answers

1. **What is the customer number for the fifth observation?**  
    The customer number for the fifth observation is **74**.

2. **What is the value difference between the lowest and fifteenth-lowest total balances?**  
The difference between the lowest and 15th-lowest total balances is **104.0**.

### Question 3

In [31]:
# Create a new column called ApplicationMonth

# Convert ApplicationDate to datetime format 
bank_loans['ApplicationDate'] = pd.to_datetime(bank_loans['ApplicationDate'], format="%d/%m/%y")

# Create a new column showing the full month name
bank_loans['ApplicationMonth'] = bank_loans['ApplicationDate'].dt.strftime("%B")

In [32]:
# Determine the number of applications in each month
monthly_application_counts = bank_loans['ApplicationMonth'].value_counts().sort_index()
print("Number of applications in each month:\n", monthly_application_counts)

Number of applications in each month:
 ApplicationMonth
April        43
August       37
December     38
February     31
January      34
July         33
June         32
March        43
May          24
November     42
October      30
September    31
Name: count, dtype: int64


In [33]:
# Determine the number of applications for each loan purpose in each month
monthly_purpose_counts = bank_loans.groupby(['ApplicationMonth', 'LoanPurpose']).size()
print("\nNumber of applications for each loan purpose in each month:\n", monthly_purpose_counts)


Number of applications for each loan purpose in each month:
 ApplicationMonth  LoanPurpose    
April             Business            4
                  Education           2
                  Furniture          11
                  New Car            11
                  Repairs             2
                                     ..
September         New Car            10
                  Other               1
                  Retraining          1
                  Small Appliance     5
                  Used Car            3
Length: 85, dtype: int64


In [34]:
# Determine the number of different loan purposes in each month
unique_purposes_per_month = bank_loans.groupby('ApplicationMonth')['LoanPurpose'].nunique()
print("\nNumber of different loan purposes in each month:\n", unique_purposes_per_month)


Number of different loan purposes in each month:
 ApplicationMonth
April        7
August       8
December     8
February     7
January      8
July         6
June         7
March        7
May          5
November     6
October      8
September    8
Name: LoanPurpose, dtype: int64


### Question 4

In [35]:
# Create a function called check_cat_censored

# Define censored keywords as a list
censored_categories = ['car', 'appliance']

# Function to check if a loan purpose belongs to a censored category
def check_cat_censored(loan_purpose):
    for keyword in censored_categories:
        if keyword.lower() in str(loan_purpose).lower():
            return True
    return False


In [36]:
# Create a new column called requires_censored
bank_loans['requires_censored'] = bank_loans['LoanPurpose'].apply(check_cat_censored)

In [37]:
# Identify customers that satisfy the two conditions at the same time
flagged_customers = bank_loans[(bank_loans['requires_censored']) & (bank_loans['MonthsCustomer'] >= 36)]

# Preview result
print("Customers with censored loans and 3+ years with the bank:")
print(flagged_customers[['CustomerNo', 'LoanPurpose', 'MonthsCustomer', 'requires_censored']].head())

Customers with censored loans and 3+ years with the bank:
    CustomerNo      LoanPurpose  MonthsCustomer  requires_censored
11          12  Small Appliance              37               True
30          31         Used Car              49               True
31          32  Small Appliance              37               True
41          42  Small Appliance              49               True
45          46  Small Appliance              37               True


### Question 5

In [38]:
# Generate a Pivot table to summarise the average age and the average number of years at current residence for different education levels and job types

# Generate pivot table grouped by EducationLevel and Job
pivot_summary = pd.pivot_table(
    bank_loans,
    index=['EducationLevel', 'Job'],
    values=['Age', 'Years'],
    aggfunc='mean'
)

# Display the result
print("Average Age and Years at Residence by Education Level and Job Type:")
print(pivot_summary)

Average Age and Years at Residence by Education Level and Job Type:
                                    Age     Years
EducationLevel    Job                            
Associate Degree  Management  39.916667  3.000000
                  Skilled     32.938776  2.673469
                  Unemployed  25.000000  1.333333
                  Unskilled   29.333333  2.666667
Bachelor's Degree Management  37.312500  3.062500
                  Skilled     32.081633  2.673469
                  Unemployed  46.666667  3.333333
                  Unskilled   36.500000  2.666667
Doctorate         Management  33.500000  2.625000
                  Skilled     35.278689  2.852459
                  Unemployed  28.000000  4.000000
                  Unskilled   36.333333  2.833333
High School       Management  44.000000  2.888889
                  Skilled     32.125000  2.729167
                  Unemployed  53.000000  4.000000
                  Unskilled   35.823529  3.176471
Master's Degree   Management  38

### Question 6

In [39]:
# Load the XML file's data into a DataFrame named mon_info
mon_info = pd.read_xml("bank_extra.xml")


In [40]:
mon_info

Mon  MaxAppNum
0   Jan         10
1   Feb         12
2   Mar         12
3   Apr         12
4   May         15
5   Jun         15
6   Jul         15
7   Aug         15
8   Sep         12
9   Oct         12
10  Nov         12
11  Dec         10

In [41]:
# Combine bank_loans and mon_info into one single DataFrame loans

from datetime import datetime
# Convert abbreviated month names in 'Mon' to full month names to match ApplicationMonth
mon_info['ApplicationMonth'] = mon_info['Mon'].apply(lambda x: datetime.strptime(x, "%b").strftime("%B"))

# Merge bank_loans with mon_info on 'ApplicationMonth'
loans = pd.merge(bank_loans, mon_info[['ApplicationMonth', 'MaxAppNum']], how='left', on='ApplicationMonth')

In [42]:
# Sort loans by the column ApplicationMonth descendingly first and then by the customer number ascendingly. Then, save the sorted results in a new DataFrame sorted_loans.
sorted_loans = loans.sort_values(by=['ApplicationMonth', 'CustomerNo'], ascending=[False, True]).reset_index(drop=True)

# Preview sorted result
print("Preview of sorted_loans DataFrame:")
print(sorted_loans[['CustomerNo', 'ApplicationMonth', 'MaxAppNum']].head())

Preview of sorted_loans DataFrame:
   CustomerNo ApplicationMonth  MaxAppNum
0           8        September         12
1           9        September         12
2          25        September         12
3          35        September         12
4          44        September         12


### Question 7

In [43]:
# Load the JSON file's data into a DataFrame named loan_info
loan_info = pd.read_json("bank_extra.json")

# The JSON is in wide format, so transpose and tidy it
loan_info = loan_info.T.reset_index()
loan_info.columns = ['LoanPurpose', 'MinDays', 'MaxDays']


In [44]:
# Combine sorted_loans and loan_info into one single DataFrame loans_final
loans_final = pd.merge(sorted_loans, loan_info, how='left', on='LoanPurpose')


In [45]:
# Sort loans_final by the customer number descendingly and save the sorted results in a new DataFrame sorted_loans_final
sorted_loans_final = loans_final.sort_values(by='CustomerNo', ascending=False).reset_index(drop=True)

# Preview result
print("Preview of sorted_loans_final DataFrame:")
print(sorted_loans_final[['CustomerNo', 'LoanPurpose', 'MinDays', 'MaxDays']].head())


Preview of sorted_loans_final DataFrame:
   CustomerNo LoanPurpose  MinDays  MaxDays
0         425    Used Car        3       14
1         424    Used Car        3       14
2         423   Furniture        1        7
3         422     New Car        3       14
4         421   Furniture        1        7
